File: 79-word-search/word-search.java
class Solution {
    public boolean exist(char[][] board, String word) {
        for(int  i=0; i<board.length; i++)
            for(int j=0; j<board[0].length; j++)
                if(board[i][j] == word.charAt(0) && valid(board, word, 0, i, j))
                    return true;

        return false;
    }

    private boolean valid(char[][] board, String word, int n, int x, int y) {
        if(n >= word.length()) return true;

        if(x >= board.length || y >= board[0].length || x < 0 || y < 0 || board[x][y] != word.charAt(n)) return false;

        boolean found;
        char temp = board[x][y];

        board[x][y] = '#';

        found = valid(board, word, n+1, x+1, y) || valid(board, word, n+1, x-1, y) ||
                valid(board, word, n+1, x, y+1) || valid(board, word, n+1, x, y-1);
        
        board[x][y] = temp;

        return found;
    }
}

File: 79-word-search/word-search.py
class Solution:
    def exist(self, board: List[List[str]], word: str) -> bool:
        rows = len(board)
        cols = len(board[0])

        def bt(idx, r, c):
            if idx == len(word):
                return True

            if r < 0 or r >= rows or c < 0 or c >= cols:
                return False

            if word[idx] != board[r][c]:
                return False

            temp = board[r][c]
            board[r][c] = "#"

            found = (
                bt(idx + 1, r - 1, c) or
                bt(idx + 1, r + 1, c) or
                bt(idx + 1, r, c - 1) or
                bt(idx + 1, r, c + 1)
            )

            board[r][c] = temp

            return found
        
        for r in range(rows):
            for c in range(cols):
                if bt(0, r, c):
                    return True
        
        return False